### ISSUE TITLE
Client performance/scalability: Multi-threaded write to one channel vs. multiple channels

### ISSUE THREAD

**Issue by mp911de** [netty/netty#3919]
Hi there, 
I'm dealing with netty client scalability with multiple connections. I maintain lettuce, a Redis client.

**TL;DR**
Why does effective performance not scale (linear) to the connection count?

**Environment/Use-Case**
- Client using Nio with shared event loop group
- Small messages (about 30 bytes) with a flush aver each message
- A small number of concurrent connections (4 connections)

**Long version**

I'm running into an issue I noticed with scalability when writing multi-threaded to one or multiple channels in netty 4.0.28. I could not find anything on SO about this topic. I found only "Hey, multithreading improves throughput" questions but mine is somehow the opposite of that.

Lettuce itself has some hot spots to be fixed. Lettuce's `writeAndFlush(...).await()` results in 60KOps/sec whereas `writeAndFlush(..., channel.voidPromise())` delivers about 1.1MOps/sec. Redis can operate as cluster (sharded cluster) so one could expect linear scaling. Effective performance drops when writing to multiple channels. 

I created a few lines of code to use pure netty (without Redis and lettuce) to check, whether lettuce is the bottleneck. My results (so far):
- Single connection writeAndFlush with await are twice as fast as multi-connection writeAndFlush
- Multi connection writeAndFlush without await are about 9 times faster than single connection writeAndFlush

I tried to use `SocketChannel` directly as a simple comparison and noticed, that the write performance is higher compared to netty. Although it's not scaling linearly it has twice ops/sec with 4 connection vs. 1 connection. Don't get me wrong, netty is gorgeous, and I'm searching how to improve performance to get higher throughput with small messages.

**Expectation**
I expected, that the throughput (from a client perspective) scales when adding connections.

The code does not have any string/byte conversion and is, therefore, faster than the results in the SO post.

SO Reference: http://stackoverflow.com/questions/31068071/netty-client-performance-scalability-multi-threaded-write-to-one-channel-vs-mu

**How to reproduce**
You can find the code at https://gist.github.com/mp911de/7a3d7206931c8c4dc581 

The code does not contain a Redis server since netty's `DiscardServer` is much more stable in performance. There are two starting points:
- JMH
- mybench.RunMain

**My JMH results**

4.0.28.Final with defaults

```
JmhTest.testMultiConnectionAwait           thrpt    5   170315,697 ±    3231,074  ops/s
JmhTest.testMultiConnectionFireAndForget   thrpt    5  9706845,564 ± 2445110,853  ops/s
JmhTest.testSingleConnectionAwait          thrpt    5   348276,047 ±   41204,225  ops/s
JmhTest.testSingleConnectionFireAndForget  thrpt    5  1071373,797 ± 1677991,227  ops/s
```

4.0.28.Final with explicit PooledByteBufAllocator

```
RunMeWithJmh.testMultiConnectionAwait           thrpt    5   167912,451 ±    5795,711  ops/s
RunMeWithJmh.testMultiConnectionFireAndForget   thrpt    5  9059877,560 ± 2765854,457  ops/s
RunMeWithJmh.testSingleConnectionAwait          thrpt    5   346812,128 ±    9397,771  ops/s
RunMeWithJmh.testSingleConnectionFireAndForget  thrpt    5  1090279,805 ± 1832319,454  ops/s
```

4.1.0.Beta5 with defaults

```
RunMeWithJmh.testMultiConnectionAwait           thrpt    5   170161,092 ±    5120,904  ops/s
RunMeWithJmh.testMultiConnectionFireAndForget   thrpt    5  4834290,916 ± 7189487,704  ops/s
RunMeWithJmh.testSingleConnectionAwait          thrpt    5   337994,099 ±   21305,377  ops/s
RunMeWithJmh.testSingleConnectionFireAndForget  thrpt    5   845681,730 ± 1317859,214  ops/s
```

4.1.0.Beta5 with explicit PooledByteBufAllocator

```
RunMeWithJmh.testMultiConnectionAwait           thrpt    5   170902,786 ±   14219,245  ops/s
RunMeWithJmh.testMultiConnectionFireAndForget   thrpt    5  4151212,491 ± 6478252,121  ops/s
RunMeWithJmh.testSingleConnectionAwait          thrpt    5   329070,549 ±   22649,903  ops/s
RunMeWithJmh.testSingleConnectionFireAndForget  thrpt    5   852931,265 ±  920164,302  ops/s
```

Hope, I did not scare you.

**Comment by normanmaurer** [comment]
@mp911de thanks for the report, and no you not scared me at least =P I'm a bit busy today but will check in detail at latest on monday.

**Comment by normanmaurer** [comment]
@mp911de could you also share the code that uses SocketChannel directly ?

**Comment by normanmaurer** [comment]
@mp911de I did some quick tests and if I use an EventLoopGroup with only one thread, using multiple connections + PooledByteBufAllocator using multiple connections out performs a single connection here.

**Comment by mp911de** [comment]
```
package mybench;

import java.io.IOException;
import java.net.SocketAddress;
import java.nio.ByteBuffer;
import java.nio.channels.SocketChannel;

import io.netty.channel.EventLoopGroup;

public class Connection {

    private SocketChannel socketChannel;

    public Connection(SocketAddress socketAddress, EventLoopGroup eventLoopGroup) throws InterruptedException {

        try {
            socketChannel = SocketChannel.open();
            socketChannel.connect(socketAddress);
            socketChannel.configureBlocking(false);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public void write(byte[] message, boolean await) {

        try {
            socketChannel.write(ByteBuffer.wrap(message));
        } catch (IOException e) {
            throw new IllegalStateException(e);
        }
    }

    public void close() {
        try {
            socketChannel.close();
        } catch (IOException e) {
            throw new IllegalStateException(e);
        }
    }
}
```

**Comment by mp911de** [comment]
A single-threaded EventLoopGroup performs about 2x faster, I got same results for single and multiple connections

**Comment by normanmaurer** [comment]
@mp911de thanks... I will check but I suspect it is how we handle writes from outside the EventLoop in Netty. If you write from outside the EventLoop we basically need to wakeup the Selector everytime you call Channel.writeAndFlush(...). This is needed as the we need to ensure we do the actual write from inside the EventLoop and so have your ChannelHandlers inbound and outbound methods called by the same Thread.

When using SocketChannel directly it will just be called from the calling thread and no wakeup needs to take place.

**Comment by mp911de** [comment]
I'll change my code to use the event loop as executor instead of the external ThreadPoolExecutor

**Comment by normanmaurer** [comment]
@mp911de let me know how it works out!

**Comment by normanmaurer** [comment]
@mp911de also another good optimation would be to not call writeAndFlush(...) everytime but call write(...) a few times and then flush. This will ensure you do a gathering write if possible and so safe sys calls.

**Comment by mp911de** [comment]
Some 15-18% improvement (from 250K to 295K ops/sec). A single-threaded EventLoopGroup is still faster.

**Comment by normanmaurer** [comment]
@mp911de you need to ensure that the channels that are contained in one Dispatcher are on the same EventLoop

**Comment by mp911de** [comment]
@normanmaurer That's my Idea from yesterday (https://github.com/mp911de/lettuce/issues/92). Using the full client achieves a boost to about 800K ops/sec instead of (100 to 150K ops/sec), but these are different numbers compared to the very slim test code, which I provided.

**Comment by mp911de** [comment]
@normanmaurer Talking about http://stackoverflow.com/questions/22559272/netty-4-channel-eventloop-mapping-in-the-nioeventloopgroup / https://github.com/netty/netty/issues/1230 ? I tried also a single-threaded EventLoopGroup per channel, but similar performance than a shared EventLoopGroup with multiple threads.

**Comment by normanmaurer** [comment]
@mp911de can we close this now with respect to #4952 ?

**Comment by mp911de** [comment]
Yes, thank you
Norman Maurer <notifications@github.com> schrieb am Di., 12. Apr. 2016 um
21:13:

**Comment by DATuan91** [comment]
@normanmaurer As author already agreed, you should close this issue
